usr/lib/common/pkey_utils.c

```diff
@@ -908,18 +908,43 @@ struct aes_xts_param {
     uint8_t param_pcc[sizeof(struct cpacf_pcc_xts_aes_256_param)];
     unsigned int fc;
     unsigned int keylen;
+    convert_key_t convert_key;
+    STDLL_TokData_t *tokdata;
+    SESSION *session;
+    OBJECT *key_obj;
 };
 
 static CK_RV pkey_aes_xts_iv_from_tweak(CK_BYTE *tweak, CK_BYTE* iv,
                                         void *cb_data)
 {
     struct aes_xts_param *param = cb_data;
-    int offset, rc;
+    convert_key_t convert_key = param->convert_key;
+    int offset, rc = 1, num_tries = 0;
+    CK_BYTE protkey[MAXPROTKEYSIZE * 2];
+    CK_ULONG protkey_len = sizeof(protkey);
+    CK_RV ret;
 
     offset = AES_XTS_PCC_I(param->keylen * 8);
     memcpy(param->param_pcc + offset, tweak, AES_BLOCK_SIZE);
 
-    rc = s390_pcc(param->fc & 0x7f, param->param_pcc);
+    while (rc != 0 && num_tries < PKEY_CONVERT_KEY_RETRIES) {
+        rc = s390_pcc(param->fc & 0x7f, param->param_pcc);
+        if (rc != 0) {
+            TRACE_DEVEL("%s rc from s390_pcc = %d, probably caused by "
+                        "an LGR. Retrying ...\n", __func__, rc);
+            ret = convert_key(param->tokdata, param->session, param->key_obj,
+                              CK_TRUE, protkey, &protkey_len);
+            if (ret != CKR_OK)
+                break;
+            /*
+             * Copy XTS key 2 into CPACF parmblock for PCC instruction. Key 2
+             * is used to create the iv from tweak via PCC. The key object
+             * itself now contains both keys in rederived form.
+             */
+            memcpy(param->param_pcc, protkey + protkey_len / 2, protkey_len / 2);
```

Comment:
I guess you should also copy the re-derived key part 1 into `param->param_km` here.

Initially both params blocks are setup with the key material in `pkey_aes_xts()` before calling `aes_xts_cipher()`.
`aes_xts_cipher()` then first calls `pkey_aes_xts_iv_from_tweak()` and after that calls `pkey_aes_xts_cipher_blocks()` multiple times.

So if the key must be re-derived inside `pkey_aes_xts_iv_from_tweak()` (due to LGR), then you must also update the key in the `param_km` so that the subsequent calls to `pkey_aes_xts_cipher_blocks()` work on the re-derived key as well.

If an LGR happens during `pkey_aes_xts_cipher_blocks()`, it is OK to only update the key in the `param_km`, since `pkey_aes_xts_iv_from_tweak()` is not called anymore.

Reply:
Yes, that makes sense. Added the memcpy for key 1 and some comment.